### PR TITLE
add zstd long range option

### PR DESCRIPTION
**Pull request by ygrek**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

add support for zstd long option for better compression of string columns to save space

Detailed description / Documentation draft:

ZSTD compression codec now takes second optional argument - window size (expressed as power of 2). If this argument is missing or specified as zero - zstd will use default window size.

### CODE REVIEW

File: src/Storages/MergeTree/MergeTreeIOSettings.h [hunk 1]
```diff
@@ -22,7 +22,7 @@ struct MergeTreeWriterSettings
     MergeTreeWriterSettings(const Settings & global_settings, bool can_use_adaptive_granularity_,
         size_t aio_threshold_, bool blocks_are_granules_size_ = false)
         : min_compress_block_size(global_settings.min_compress_block_size)
-        , max_compress_block_size(global_settings.min_compress_block_size)
+        , max_compress_block_size(global_settings.max_compress_block_size)
```
Comment (alexey-milovidov) [inline src/Storages/MergeTree/MergeTreeIOSettings.h:25]: :+1:

File: src/Core/Settings.h [hunk 2]
```diff
@@ -33,7 +33,7 @@ class IColumn;
 
 #define COMMON_SETTINGS(M) \
     M(UInt64, min_compress_block_size, 65536, "The actual size of the block to compress, if the uncompressed data less than max_compress_block_size is no less than this value and no less than the volume of data for one mark.", 0) \
-    M(UInt64, max_compress_block_size, 1048576, "The maximum size of blocks of uncompressed data before compressing for writing to a table.", 0) \
+    M(UInt64, max_compress_block_size, 8048576, "The maximum size of blocks of uncompressed data before compressing for writing to a table.", 0) \
```
Comment (alexey-milovidov) [inline src/Core/Settings.h:36]: We should not change it by default, because it will slow down point queries.

**Comment by ygrek** [comment]
currently failing fast tests are 
01200_mutations_memory_consumption (which also fails on master) with
```
01200_mutations_memory_consumption:                                     [ FAIL ] - return code 60
[devbox1] 2020.12.08 23:40:04.368173 [ 117234 ] {8c8e0ab5-dc07-49bf-9f44-095afdc0bfc0} <Error> executeQuery: Code: 60, e.displayText() = DB::Exception: Table system.part_log doesn't exist. (version 20.12.1.1) (from 127.0.0.1:40216) (in query: SELECT DISTINCT read_bytes >= peak_memory_usage FROM system.part_log WHERE event_type = 'MutatePart' AND table = 'table_with_single_pk' AND database = currentDatabase(); ), Stack trace (when copying this message, always include the lines below):
```
and
01593_concurrent_alter_mutations_kill_many_replicas which I cannot reproduce : 
```
$ ./tests/clickhouse-test 01593_concurrent_alter_mutations_kill_many_replicas
Using queries from '/home/user/clickhouse/tests/queries' directory
Connecting to ClickHouse server... OK
Won't run stateful tests because test data wasn't loaded.
No tests were run.
```

**Comment by abyss7** [comment]
Is it ready for review? Can you fix the title and description then.

**Comment by ygrek** [comment]
updated

(I was trying to double-check that it works on latest tag, but fail to build it after rebase so far)

File: src/Compression/CompressionCodecZSTD.cpp [hunk 3]
```diff
@@ -79,9 +93,31 @@ void registerCodecZSTD(CompressionCodecFactory & factory)
...
             level = literal->value.safeGet<UInt64>();
             if (level > ZSTD_maxCLevel())
-                throw Exception("ZSTD codec can't have level more that " + toString(ZSTD_maxCLevel()) + ", given " + toString(level), ErrorCodes::ILLEGAL_CODEC_PARAMETER);
+                throw Exception(
+                    "ZSTD codec can't have level more that " + toString(ZSTD_maxCLevel()) + ", given " + toString(level),
+                    ErrorCodes::ILLEGAL_CODEC_PARAMETER);
+            if (arguments->children.size() > 1)
+            {
+                const auto * window_literal = children[1]->as<ASTLiteral>();
+                if (!window_literal)
```
Comment (abyss7) [inline src/Compression/CompressionCodecZSTD.cpp:102]: You're not checking if this is really an integer at this point

File: src/Compression/CompressionCodecZSTD.cpp [hunk 4]
```diff
@@ -79,9 +93,31 @@ void registerCodecZSTD(CompressionCodecFactory & factory)
 
             level = literal->value.safeGet<UInt64>();
             if (level > ZSTD_maxCLevel())
-                throw Exception("ZSTD codec can't have level more that " + toString(ZSTD_maxCLevel()) + ", given " + toString(level), ErrorCodes::ILLEGAL_CODEC_PARAMETER);
+                throw Exception(
+                    "ZSTD codec can't have level more that " + toString(ZSTD_maxCLevel()) + ", given " + toString(level),
```
Comment (abyss7) [inline src/Compression/CompressionCodecZSTD.cpp:97]: typo `that` → `than`

File: src/Compression/CompressionCodecZSTD.cpp [hunk 5]
```diff
@@ -79,9 +93,31 @@ void registerCodecZSTD(CompressionCodecFactory & factory)
...
+                const int window_log = window_literal->value.safeGet<UInt64>();
+
+                ZSTD_bounds window_log_bounds = ZSTD_cParam_getBounds(ZSTD_c_windowLog);
+                if (ZSTD_isError(window_log_bounds.error))
+                    throw Exception(
+                        "ZSTD windowLog paramater is not supported " + std::string(ZSTD_getErrorName(window_log_bounds.error)),
+                        ErrorCodes::ILLEGAL_CODEC_PARAMETER);
+                if (window_log != 0 && (window_log > window_log_bounds.upperBound || window_log < window_log_bounds.lowerBound))
+                    throw Exception(
+                        "ZSTD codec can't have window log more that " + toString(window_log_bounds.upperBound) + " and lower than "
```
Comment (abyss7) [inline src/Compression/CompressionCodecZSTD.cpp:114]: Another typo

File: src/Compression/CompressionCodecZSTD.cpp [hunk 6]
```diff
@@ -79,9 +93,31 @@ void registerCodecZSTD(CompressionCodecFactory & factory)
...
+                    throw Exception("ZSTD codec second argument must be integer", ErrorCodes::ILLEGAL_CODEC_PARAMETER);
+
+                const int window_log = window_literal->value.safeGet<UInt64>();
+
+                ZSTD_bounds window_log_bounds = ZSTD_cParam_getBounds(ZSTD_c_windowLog);
+                if (ZSTD_isError(window_log_bounds.error))
+                    throw Exception(
+                        "ZSTD windowLog paramater is not supported " + std::string(ZSTD_getErrorName(window_log_bounds.error)),
+                        ErrorCodes::ILLEGAL_CODEC_PARAMETER);
+                if (window_log != 0 && (window_log > window_log_bounds.upperBound || window_log < window_log_bounds.lowerBound))
```
Comment (abyss7) [inline src/Compression/CompressionCodecZSTD.cpp:112]: Why should the `0` be special value for boundary check?

**Review by abyss7 — CHANGES_REQUESTED** [review]
The code is good, but there is a couple of minor issues. The main request is to add a simple test that will cover your new code at least: like reading and writing with custom window log param.

File: src/Compression/CompressionCodecZSTD.cpp (hunk 3 above)
Comment (ygrek) [reply src/Compression/CompressionCodecZSTD.cpp:102]: hm, this code does the same as other compression codecs (and same what is done for first parameters in zstd codec, above)

File: src/Compression/CompressionCodecZSTD.cpp (hunk 6 above)
Comment (ygrek) [reply src/Compression/CompressionCodecZSTD.cpp:112]: value 0 means 'use default' in libzstd, i will add comment

**Comment by ygrek** [comment]
test added

**Review by abyss7 — APPROVED** [review]

**Pull request merged**